Backtracking doesn't find last event when no progress

https://github.com/akka/akka-persistence-r2dbc/runs/4079463405?check_suite_focus=true#step:7:2415

```
[2021-11-02 11:49:02,221] [INFO] [akka.actor.testkit.typed.scaladsl.LogCapturing] [] [] [pool-1-thread-1-ScalaTest-running-EndToEndSpec] - Logging finished for test [akka.projection.r2dbc.EndToEndSpec: A R2DBC projection with eventsBySlices source must handle all events exactlyOnce] that [Failed(java.lang.AssertionError: timeout (20 seconds) while expecting 200 messages (got 194))]
2414
<-- [akka.projection.r2dbc.EndToEndSpec: A R2DBC projection with eventsBySlices source must handle all events exactlyOnce] End of log messages of test that [Failed(java.lang.AssertionError: timeout (20 seconds) while expecting 200 messages (got 194))]
2415
[info] - must handle all events exactlyOnce *** FAILED *** (26 seconds, 502 milliseconds)
2416
[info]   java.lang.AssertionError: timeout (20 seconds) while expecting 200 messages (got 194)
2417
[info]   at akka.actor.testkit.typed.internal.TestProbeImpl.assertFail(TestProbeImpl.scala:399)
2418
[info]   at akka.actor.testkit.typed.internal.TestProbeImpl.$anonfun$receiveMessages_internal$1(TestProbeImpl.scala:262)
2419
[info]   at akka.actor.testkit.typed.internal.TestProbeImpl.$anonfun$receiveMessages_internal$1$adapted(TestProbeImpl.scala:257)
2420
[info]   at scala.collection.immutable.Range.map(Range.scala:59)
2421
[info]   at akka.actor.testkit.typed.internal.TestProbeImpl.receiveMessages_internal(TestProbeImpl.scala:257)
2422
[info]   at akka.actor.testkit.typed.internal.TestProbeImpl.receiveMessages(TestProbeImpl.scala:247)
2423
[info]   at akka.projection.r2dbc.EndToEndSpec.$anonfun$new$2(EndToEndSpec.scala:217)
```

--- a/core/src/main/scala/akka/persistence/r2dbc/internal/BySliceQuery.scala
+++ b/core/src/main/scala/akka/persistence/r2dbc/internal/BySliceQuery.scala
@@ -48,7 +48,7 @@ import org.slf4j.Logger
       if (backtracking) latestBacktracking.timestamp
       else latest.timestamp
 
-    def nextQueryUntilTimestamp: Option[Instant] =
+    def nextQueryToTimestamp: Option[Instant] =
       if (backtracking) Some(latest.timestamp)
       else None
   }
@@ -68,7 +68,7 @@ import org.slf4j.Logger
         minSlice: Int,
         maxSlice: Int,
         fromTimestamp: Instant,
-        untilTimestamp: Option[Instant],
+        toTimestamp: Option[Instant],
         behindCurrentTime: FiniteDuration,
         backtracking: Boolean): Source[SerializedRow, NotUsed]
   }
@@ -125,7 +125,7 @@ import org.slf4j.Logger
               minSlice,
               maxSlice,
               state.latest.timestamp,
-              untilTimestamp = Some(toDbTimestamp),
+              toTimestamp = Some(toDbTimestamp),
               behindCurrentTime = Duration.Zero,
               backtracking = false)
             .via(deserializeAndAddOffset(state.latest)))
@@ -272,7 +272,7 @@ import org.slf4j.Logger
             minSlice,
             maxSlice,
             newState.nextQueryFromTimestamp,
-            newState.nextQueryUntilTimestamp,
+            newState.nextQueryToTimestamp,
             behindCurrentTime,
             backtracking = newState.backtracking)
           .via(deserializeAndAddOffset(newState.currentOffset)))

--- a/core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/QueryDao.scala
+++ b/core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/QueryDao.scala
@@ -46,7 +46,7 @@ private[r2dbc] class QueryDao(settings: R2dbcSettings, connectionFactory: Connec
     "SELECT transaction_timestamp() AS db_timestamp"
 
   private def eventsBySlicesRangeSql(
-      maxDbTimestampParam: Boolean,
+      toDbTimestampParam: Boolean,
       behindCurrentTime: FiniteDuration,
       backtracking: Boolean): String = {
     var p = 0
@@ -56,8 +56,8 @@ private[r2dbc] class QueryDao(settings: R2dbcSettings, connectionFactory: Connec
       "$" + p
     }
 
-    def maxDbTimestampParamCondition =
-      if (maxDbTimestampParam) s"AND db_timestamp < ${nextParam()}" else ""
+    def toDbTimestampParamCondition =
+      if (toDbTimestampParam) s"AND db_timestamp <= ${nextParam()}" else ""
 
     def behindCurrentTimeIntervalCondition =
       if (behindCurrentTime > Duration.Zero)
@@ -75,7 +75,7 @@ private[r2dbc] class QueryDao(settings: R2dbcSettings, connectionFactory: Connec
     s"FROM $journalTable " +
     s"WHERE entity_type = ${nextParam()} " +
     s"AND slice BETWEEN ${nextParam()} AND ${nextParam()} " +
-    s"AND db_timestamp >= ${nextParam()} $maxDbTimestampParamCondition $behindCurrentTimeIntervalCondition " +
+    s"AND db_timestamp >= ${nextParam()} $toDbTimestampParamCondition $behindCurrentTimeIntervalCondition " +
     s"AND deleted = false " +
     s"ORDER BY db_timestamp, seq_nr " +
     s"LIMIT ${nextParam()}"
@@ -122,19 +122,19 @@ private[r2dbc] class QueryDao(settings: R2dbcSettings, connectionFactory: Connec
       minSlice: Int,
       maxSlice: Int,
       fromTimestamp: Instant,
-      untilTimestamp: Option[Instant],
+      toTimestamp: Option[Instant],
       behindCurrentTime: FiniteDuration,
       backtracking: Boolean): Source[SerializedJournalRow, NotUsed] = {
     val result = r2dbcExecutor.select(s"select eventsBySlices [$minSlice - $maxSlice]")(
       connection => {
         val stmt = connection
           .createStatement(
-            eventsBySlicesRangeSql(maxDbTimestampParam = untilTimestamp.isDefined, behindCurrentTime, backtracking))
+            eventsBySlicesRangeSql(toDbTimestampParam = toTimestamp.isDefined, behindCurrentTime, backtracking))
           .bind(0, entityType)
           .bind(1, minSlice)
           .bind(2, maxSlice)
           .bind(3, fromTimestamp)
-        untilTimestamp match {
+        toTimestamp match {
           case Some(until) =>
             stmt.bind(4, until)
             stmt.bind(5, settings.querySettings.bufferSize)

--- a/core/src/main/scala/akka/persistence/r2dbc/state/scaladsl/DurableStateDao.scala
+++ b/core/src/main/scala/akka/persistence/r2dbc/state/scaladsl/DurableStateDao.scala
@@ -237,19 +237,19 @@ private[r2dbc] class DurableStateDao(settings: R2dbcSettings, connectionFactory:
       minSlice: Int,
       maxSlice: Int,
       fromTimestamp: Instant,
-      untilTimestamp: Option[Instant],
+      toTimestamp: Option[Instant],
       behindCurrentTime: FiniteDuration,
       backtracking: Boolean): Source[SerializedStateRow, NotUsed] = {
     val result = r2dbcExecutor.select(s"select stateBySlices [$minSlice - $maxSlice]")(
       connection => {
         val stmt = connection
           .createStatement(
-            stateBySlicesRangeSql(maxDbTimestampParam = untilTimestamp.isDefined, behindCurrentTime, backtracking))
+            stateBySlicesRangeSql(maxDbTimestampParam = toTimestamp.isDefined, behindCurrentTime, backtracking))
           .bind(0, entityType)
           .bind(1, minSlice)
           .bind(2, maxSlice)
           .bind(3, fromTimestamp)
-        untilTimestamp match {
+        toTimestamp match {
           case Some(until) =>
             stmt.bind(4, until)
             stmt.bind(5, settings.querySettings.bufferSize)

--- a/core/src/test/scala/akka/persistence/r2dbc/query/EventsBySliceBacktrackingSpec.scala
+++ b/core/src/test/scala/akka/persistence/r2dbc/query/EventsBySliceBacktrackingSpec.scala
@@ -120,10 +120,6 @@ class EventsBySliceBacktrackingSpec
       // no backtracking yet
       result.expectNoMessage(settings.querySettings.refreshInterval + 100.millis)
 
-      // FIXME one case that isn't covered yet is if there is no progress (no more events) then
-      // there will not be any new backtracking query and missed events will not be found (until there are
-      // new events that move the latest offset forward)
-
       // after 1/2 of the backtracking widow, to kick off a backtracking query
       writeEvent(
         slice1,

--- a/core/src/test/scala/akka/persistence/r2dbc/query/EventsBySliceBacktrackingSpec.scala
+++ b/core/src/test/scala/akka/persistence/r2dbc/query/EventsBySliceBacktrackingSpec.scala
@@ -107,12 +107,18 @@ class EventsBySliceBacktrackingSpec
       env3.eventOption shouldBe None
       // but it can be lazy loaded
       query.loadEnvelope[String](env3.persistenceId, env3.sequenceNr).futureValue.get.eventOption shouldBe Some("e1-1")
+      // backtracking up to (and equal to) the same offset
+      val env4 = result.expectNext()
+      env4.persistenceId shouldBe pid1
+      env4.sequenceNr shouldBe 2L
+      env4.eventOption shouldBe None
+
       result.expectNoMessage(100.millis) // not e1-2
 
       writeEvent(slice1, pid1, 3L, startTime.plusMillis(3), "e1-3")
-      val env4 = result.expectNext()
-      env4.persistenceId shouldBe pid1
-      env4.sequenceNr shouldBe 3L
+      val env5 = result.expectNext()
+      env5.persistenceId shouldBe pid1
+      env5.sequenceNr shouldBe 3L
 
       // before e1-3 so it will not be found by the normal query
       writeEvent(slice2, pid2, 1L, startTime.plusMillis(2), "e2-1")
@@ -127,16 +133,12 @@ class EventsBySliceBacktrackingSpec
         4L,
         startTime.plusMillis(settings.querySettings.backtrackingWindow.toMillis / 2).plusMillis(4),
         "e1-4")
-      val env5 = result.expectNext()
-      env5.persistenceId shouldBe pid1
-      env5.sequenceNr shouldBe 4L
-
-      // backtracking finds it,  and it also emits duplicates (by design)
-      // e1-1 was already handled by previous backtracking query
       val env6 = result.expectNext()
       env6.persistenceId shouldBe pid1
-      env6.sequenceNr shouldBe 2L
+      env6.sequenceNr shouldBe 4L
 
+      // backtracking finds it,  and it also emits duplicates (by design)
+      // e1-1 and e1-2 were already handled by previous backtracking query
       val env7 = result.expectNext()
       env7.persistenceId shouldBe pid2
       env7.sequenceNr shouldBe 1L
@@ -144,8 +146,10 @@ class EventsBySliceBacktrackingSpec
       val env8 = result.expectNext()
       env8.persistenceId shouldBe pid1
       env8.sequenceNr shouldBe 3L
-      result.expectNoMessage(100.millis) // not e1-4
-      result.expectNoMessage(settings.querySettings.refreshInterval)
+
+      val env9 = result.expectNext()
+      env9.persistenceId shouldBe pid1
+      env9.sequenceNr shouldBe 4L
     }
   }
 

--- a/projection/src/test/scala/akka/projection/r2dbc/EventSourcedEndToEndSpec.scala
+++ b/projection/src/test/scala/akka/projection/r2dbc/EventSourcedEndToEndSpec.scala
@@ -310,13 +310,9 @@ class EventSourcedEndToEndSpec
 
       // pid3, seqNr 8 is missing (knows 7) when receiving 9
       writeEvent(pid3, 9L, startTime.plusMillis(4), "e3-9")
-      processedProbe.expectNoMessage(journalSettings.querySettings.refreshInterval + 500.millis)
+      processedProbe.expectNoMessage(journalSettings.querySettings.refreshInterval + 2000.millis)
 
-      // but backtracking can fill in the gaps
-      // need some progress because backtracking will not exceed the latest offset
-      writeEvent(pid1, 2L, startTime.plusMillis(5), "e1-2")
-      processedProbe.receiveMessage().envelope.event shouldBe "e1-2"
-      // backtracking will pick up pid3 seqNr 8 and 9
+      // but backtracking can fill in the gaps, backtracking will pick up pid3 seqNr 8 and 9
       writeEvent(pid3, 8L, startTime.plusMillis(3), "e3-8")
       val possibleDelay =
         journalSettings.querySettings.backtrackingBehindCurrentTime + journalSettings.querySettings.refreshInterval + processedProbe.remainingOrDefault

